### PR TITLE
[Compiler-v2] Relax check on whether a local can be written

### LINKED ISSUE #12888: [Bug][compiler-v2] false positive `cannot assign to borrowed` error
# 🐛 Bug

When compiling the following code using V2:

```
module 0x8675309::M {

    fun foo(x: &vector<u64>): (vector<u64>, u64) {
        (*x, 0)
    }

    fun test() {
        let y = vector[];
        (y, _) = foo(&y);
    }

}
```

The following error is generated:
```
error: cannot assign to borrowed local `y`
   ┌─ tests/reference-safety/debug.move:39:18
   │
39 │         (y, _) = foo(&y);
   │                  ^^^^^^^
   │                  │   │
   │                  │   previous local borrow
   │                  attempted to assign here
```

While V1 does not generate the error.

### PR DESCRIPTION
**Pull request by rahxephon89**
## Description

Previously, as long as the temp is borrowed, it cannot be written. This PR relaxes the condition that if all borrowing temps are not alive after the statement, it can be written.

close #12888 

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Existing test cases and newly added test cases

## Key Areas to Review

Whether this change can be applied to all checks on local writes.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-compiler-v2/tests/reference-safety/write_ref_dest.exp [hunk 1]
```diff
@@ -0,0 +1,2 @@
+
+============ bytecode verification succeeded ========
```
Comment (brmataptos) [inline third_party/move/move-compiler-v2/tests/reference-safety/write_ref_dest.exp:2]: Add this to transactional tests with paranoid mode enabled.

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/tests/reference-safety/write_ref_dest.exp:2]: done

**Review by brmataptos — APPROVED** [review]
Looks ok to me.

File: third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs [hunk 2]
```diff
@@ -1005,22 +1015,32 @@ impl<'env, 'state> LifetimeAnalysisStep<'env, 'state> {
...
-            )
+            let children = self.state.transitive_children(label);
+            // cannot be assigned if the destination is currently borrowed by temps that are alive
+            // after the statement
+            if self
+                .state
+                .leaves_for_labels(&children)
+                .iter()
+                .any(|temp| self.alive.after.contains_key(temp) && *temp != local)
+            {
```
Comment (wrwg) [inline third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:1034]: This looks sound, however, I wonder why this is needed. There seems to be a more deeper bug which triggers this treatment.

In general, the invariant holds that if something is in the `temp_to_label` map, its also directly or indirectly alive. 

So why is then `local` in the map, in the case none of the children are?

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:1034]: The original looks like this:
```
...
  1: $t2 := borrow_local($t0)
  2: ($t0, $t1) := M::foo($t2)
...  
```
$2 is borrowed from $0 but $0 is reassigned in this function call so the current check will raise an error. In this revised check, we first try to find if there are any temps borrowing local after this program point. If not, we don't generate an error for it.

Comment (wrwg) [reply third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs:1034]: The issue is that t2 is going out of scope in the same step where it is used, so at the moment we assign t0, it is actually not used any more. Your fix is not the right way to deal with this, rather we want to fix this more general. One thing we could do is pull the cleanup we do after and before each instruction in `execute` into its own function and call that function before we make the check for the assigned destination. We'd want to do the same for other places were we assign variables.

Comment (rahxephon89) [reply third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs]: Added the logic to release references.

**Review by wrwg — APPROVED** [review]

**Pull request merged**